lib/1.2/dml12-compatibility.dml

```diff
@@ -63,40 +63,131 @@ template write_only_attr is pseudo_attr {
 // this template can be instantiated to make the override effective when
 // imported from a DML 1.2 device.
 template dml12_compat_io_memory_access is bank {
-    method access(generic_transaction_t *memop, physical_address_t offset,
+    param _have_dml12_compat_io_memory_access = true;
+    method access(generic_transaction_t *mop, physical_address_t offset,
                   physical_address_t size) throws {
-        if (!this.io_memory_access(memop, offset, NULL))
+        if (!this.io_memory_access(mop, offset, NULL))
             throw;
     }
 }
 
-in each bank {
+template dml12_compat_access_override {
+    // 'in each':ed in all banks to permit calls to the otherwise DML
+    // 1.4-specific functions io_memory_access and transaction_access.
+    // Can be instantiated explicitly to avoid rank ambiguities.
     method io_memory_access(generic_transaction_t *memop, uint64 offset,
                             void *aux) -> (bool) default {
-        local uint64 size = SIM_get_mem_op_size(memop);
+        log info, 4: "entering compatibility wrapper"
+            + " for io_memory_access overrides";
+        local atom_t list[7];
+        if (memop->transaction == NULL) {
+            log info, 4: "io_memory_access: wrapping memop in transaction";
+            local mem_op_type_t type = memop->type;
+
+            local transaction_flags_t flags = 0
+                | ((type & Sim_Trn_Instr) == 0 ? 0 : Sim_Transaction_Fetch)
+                | ((type & Sim_Trn_Write) == 0 ? 0 : Sim_Transaction_Write)
+                | ((type & Sim_Trn_Control) == 0 ? 0
+                   : Sim_Transaction_Control)
+                | (memop->inquiry ? Sim_Transaction_Inquiry : 0)
+                | (memop->atomic ? Sim_Transaction_Atomic : 0)
+                | (memop->non_coherent ? Sim_Transaction_Incoherent : 0);
+
+            list = {
+                ATOM_completion(NULL),
+                ATOM_flags(flags),
+                ATOM_size(memop->size),
+                ATOM_data(memop->real_address),
+                ATOM_initiator(memop->ini_ptr),
+                ATOM_memop(memop),
+                ATOM_list_end(0),
+            };
+
+            if (memop->inverse_endian) {
+                // reverse endian is hard to fully translate outside core,
+                // but this should not be a problem in practice
+                log unimpl: "transaction support for inverse endian memops";
+            }
+        } else {
+            log info, 4: "Reusing memop's existing transaction";
+            // still need to chain the transaction, to ensure
+            // transaction_access will get the memop back untouched
+            // through ATOM_get_transaction_memop
+            list = {
+                ATOM_memop(memop),
+                ATOM_list_end(0),
+                ATOM_list_end(0),
+                ATOM_list_end(0),
+                ATOM_list_end(0),
+                ATOM_list_end(0),
+                ATOM_list_end(0),
```

Comment:
This is silly. Do you need every element beyond the first `ATOM_list_end` to have determinate value? Why not:
```C
list[0] = ATOM_memop(memop);
list[1] = ATOM_list_end(0);
```
And leave it at that?

Reply:
good catch